experimental/cache_size.cc

```diff
@@ -0,0 +1,65 @@
+#include <stdlib.h>
+#include <time.h>
+
+#include <algorithm>
+#include <iostream>
+
+int64_t MAX_SIZE = 16 * 1024 * 1024;
+int64_t MIN_SIZE = 1024;
+int64_t ACCESSES_NUM = 64 * 1024 * 1024;
+int REPEATS = 50;
+
+clock_t whack_cache(const int64_t sz, int64_t* accesses) {
+  char* buf = new char[sz];
+
+  clock_t start = clock();
+  // TODO fencing
+  for (int64_t i = 0; i < ACCESSES_NUM; i++) {
+    ++buf[(accesses[i] * 64) % sz];
+  }
+  // TODO: elapsed time generates zero
+  clock_t elapsed = clock() - start;
+
+  double cpu_time_used = ((double)elapsed) / CLOCKS_PER_SEC;
+
+  delete[] buf;
```

Comment:
There's actually a really great way to do this automatically using "RAII" :D

In C++, you can allocate a zero-initialized array of characters using one of the container types on the heap, e.g. [`std::string`](https://en.cppreference.com/w/cpp/string/basic_string) or [`std::vector<char>`](https://en.cppreference.com/w/cpp/container/vector)

(There's a few more options as well, but keeping it simple for now.)

So instead of:

```c++
char* buf = new char[sz];
...
delete[] buf;
```

You could write:

```c++
std::vector<char> buf(sz);
...
```

With no `delete []` call necessary -- it will automatically destruct the vector at the end of the scope.

In general `std::vector` won't always work, and you can store objects of any type in a [`std::unique_ptr`](https://en.cppreference.com/w/cpp/memory/unique_ptr) -- it's just a pointer container that automatically invokes delete. So another way to go would be to store buf inside a `unique_ptr<char[]>`, but in this case, a vector is easier.

---

(For what it's worth, the original `char* buf = new char[sz];` I think has a bug. I believe it doesn't initialize the array, so line 18 is undefined behavior. This is not present in the version using vector above, and can be removed in the use of `new` with value-initialization: it should instead be `new char[sz]()`. See https://stackoverflow.com/questions/2204176/how-to-initialise-memory-with-new-operator-in-c)

Reply:
`std::vector` can have a capacity higher than its size, so not sure if that can be an issue here.

Reply:
I was hesitant to use `std::vector` as "the storage of the vector is handled automatically", so I thought maybe the size of vector can become different than the size mentioned in code. 
As @sivachandra also mentioned I am not sure whether it can pose a problem here or not.

Reply:
The capacity can be different than the size (i.e. it can allocate more space than it actually uses), but I don't think that should change anything. (There is always space to the side of the allocation -- if anything, a higher capacity is better since it guarantees nobody else will touch neighboring memory).

That said, maybe it's bad to use vector after all, since it can store on the _stack_, which would interfere with things.

So as an alternative, how about instead having:

```c++
std::unique_ptr<char[]> buf(new char[sz]());
...
// no delete[] call because unique_ptr deletes automatically
```

I think this should be just as easy as vector, but get rid of all the concerns we've mentioned so far:

1) memory is managed automatically, no need for manual delete
2) capacity == size -- it's literally the same kind of array stuff, except with the `delete[]` done automatically
3) by using `new char[n]()` instead of `new char[n]`, we zero-initialize the array, removing the undefined behavior.